src/test/java/org/elasticsearch/cluster/metadata/MetaDataTests.java

```diff
@@ -579,6 +592,109 @@ public void concreteIndicesWildcardExpansion() {
         assertThat(newHashSet(md.concreteIndices(IndicesOptions.fromOptions(true, true, true, true), "testX*")), equalTo(newHashSet("testXXX", "testXXY", "testXYY")));
     }
 
+    /**
+     * test resolving _all pattern (null, empty array or "_all") for random IndicesOptions
+     */
+    @Test
+    public void testConcreteIndicesAllPatternRandom() {
+        for (int i = 0; i < 10; i++) {
+            String[] allIndices = null;
+            switch (randomIntBetween(0, 2)) {
+            case 0:
+                break;
+            case 1:
+                allIndices = new String[0];
+                break;
+            case 2:
+                allIndices = new String[] { MetaData.ALL };
+                break;
+            }
+
+            IndicesOptions indicesOptions = IndicesOptions.fromOptions(randomBoolean(), randomBoolean(), randomBoolean(), randomBoolean());
+            MetaData metadata = MetaData.builder().build();
+
+            // with no indices, asking for all indices should return empty list or exception, depending on indices options
+            if (indicesOptions.allowNoIndices()) {
+                String[] concreteIndices = metadata.concreteIndices(indicesOptions, allIndices);
+                assertThat(concreteIndices, notNullValue());
+                assertThat(concreteIndices.length, equalTo(0));
+            } else {
+                try {
+                    metadata.concreteIndices(indicesOptions, allIndices);
+                    fail("calling concreteIndices for _all indices with allowNoIndices=false should throw exception");
+                } catch (ElasticsearchIllegalArgumentException | IndexMissingException e) {
```

Comment:
can we separate the catch with multiple exceptions to clarify what is expected when?